development/extensions/tutorial_php_events_and_extending_extensions.rst

```diff
@@ -0,0 +1,344 @@
+core files, and getting class names based on files it finds). The return of the
+``find()`` method is an array of file paths that match the given criteria.
+
+For example, assume you have an extension and you need to get a collection of all
+PNG image files from one of its directories. You could use the extension manager
+to load the finder, and traverse the extension's image directory as follows:
+
+.. code-block:: php
+
+  $finder = $extension_manager->get_finder();
+
+  $images = $finder
+      ->extension_suffix('.png')
+      ->extension_directory('/images')
+      ->find_from_extension('demo', $phpbb_root_path . 'ext/acme/demo/');
+
+The ``$images`` array would look something like (as you can see the image paths are
+contained in the array keys):
+
+.. code-block:: php
+
+  array(
+      'ext/acme/demo/images/image1.png' => 'demo',
+      'ext/acme/demo/images/image2.png' => 'demo',
+  );
+
+.. note::
+  The method ``find_from_extension`` used above will only search in that specific
+  extension. If you want to search for images in all extensions, you use ``suffix``,
+  ``directory`` and ``find`` instead:
+
+  .. code-block:: php
+
+    $finder = $extension_manager->get_finder();
+
+    $images = $finder
+        ->find('.png')
+        ->directory('/images')
+        ->find($phpbb_root_path . 'ext/');
+
+Using service replacement
+=========================
+.. warning::
+  Only use service replacements if your extensions needs to be compatible with
+  phpBB 3.1. For 3.2 and later, see `Using service decoration`_.
+
+.. warning::
+  You can't assume the order of a service is defined in phpBB if other extensions
+  are installed. Be really carefully when using service replacements.
+
+With a service replacement you can replace an existing service in phpBB (or in an
+extension) with your own service. Your replacement should type match the original
+service (so if the original service implements an interface, you should at least
+implement that specific interface. If it is a concrete class, you will need to
+extend that class). The best way to accomplish this is by extending the original
+class, and only replace the features you want to change.
+
+To replace a core phpBB service, you simply name your extension's service name
+with the same name as the service in phpBB's core.
+
+For example, to replace the config, replace the ``config`` service in phpBB with your own
+implementation, your service configuration would look like this:
+
+.. code-block:: yaml
+
+  config:
+      class: acme\demo\config\db
+      arguments:
+          - '@auth'
+          - '@passwords.manager'
+          - '@acme.demo.db_reader'
+
+Your PHP class should extend ``\phpbb\config\db`` which is original class in
+phpBB. To avoid any unforseen problems, you should not extend directly
+from ``\phpbb\config\config``. If both implemented the same interface, you could
+have extended from ``\phpbb\config\config``.
```

Comment:
I'm not sure to understand what you want to say here.

In order to replace the config service, the class must extends `\phpbb\config\config` or any of its subclasses. That's the contract here (but not easy to guess as there isn't any interface...). So considering that there isn't any interface I think your statement is correct.